### PR TITLE
matrix-sdk: eyre feature for event handler

### LINKED ISSUE #324: `eyre` feature for event handler error handling
Currently there's the `anyhow` feature that allows returning `anyhow::Error`s from event handlers. Implemented here

https://github.com/matrix-org/matrix-rust-sdk/blob/b94faa3121c987a568b531e7d5e5bd033713e9b3/matrix_sdk/src/event_handler.rs#L192-L199

Additionally an `eyre` feature would be helpful to allow returning `eyre::Report`.

Follow-up of https://github.com/matrix-org/matrix-rust-sdk/issues/322#issuecomment-916247273

### LINKED ISSUE #347: `anyhow` feature breaks compilation
**Describe the bug**

```rust
error[E0119]: conflicting implementations of trait `event_handler::EventHandlerResult` for type `std::result::Result<(), anyhow::Error>`
   --> crates/matrix-sdk/src/event_handler.rs:200:1
    |
191 | impl<E: std::error::Error> EventHandlerResult for Result<(), E> {
    | --------------------------------------------------------------- first implementation here
...
200 | impl EventHandlerResult for anyhow::Result<()> {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `std::result::Result<(), anyhow::Error>`
    |
    = note: upstream crates may add a new impl of trait `std::error::Error` for type `anyhow::Error` in future versions
```

**To Reproduce**
1. Enable `anyhow` feature
2. Compile

**Expected behavior**
Should compile

### PR DESCRIPTION
**Pull request by johannescpk**
Fixes #347 
Fixes #324

### CODE REVIEW

File: crates/matrix-sdk/src/event_handler.rs [hunk 1]
```diff
@@ -197,6 +198,7 @@ impl<E: std::error::Error> EventHandlerResult for Result<(), E> {
 }
 
 #[cfg(feature = "anyhow")]
+#[cfg(not(feature = "eyre"))]
```
Comment (poljar) [inline crates/matrix-sdk/src/event_handler.rs:201]: Is that `not` here needed? This means that you can't enable both features, or in other words if you enable both, both get disabled instead?

**Review by poljar — COMMENTED** [review]
I think @jplatte had some other idea how to handle the generic implementation for this so we'll wait up and see what he has to say.

File: crates/matrix-sdk/src/event_handler.rs [hunk 2]
```diff
@@ -188,6 +188,7 @@ impl EventHandlerResult for () {
     fn print_error(&self, _event_type: &str) {}
 }
 
+#[cfg(not(any(feature = "anyhow", feature = "eyre")))]
```
Comment (jplatte) [inline crates/matrix-sdk/src/event_handler.rs:191]: This means other error types no longer work with either feature activated. That seems like a very bad idea. I'm working on an alternative fix.

**Comment by johannescpk** [comment]
Superseded by https://github.com/matrix-org/matrix-rust-sdk/pull/352

**Pull request closed without merging**